server/src/main/java/org/opensearch/cluster/service/MasterTaskThrottler.java

```diff
@@ -0,0 +1,153 @@
+
+    public static final Setting<Settings> THRESHOLD_SETTINGS = Setting.groupSetting(
+        "master.throttling.thresholds.",
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * To configure more task for throttling, override getMasterThrottlingKey method with task name in task executor.
+     * Verify that throttled tasks would be retry.
+     *
+     * Added retry mechanism in TransportMasterNodeAction so it would be retried for customer generated tasks.
+     */
+    public static Set<String> CONFIGURED_TASK_FOR_THROTTLING = Collections.unmodifiableSet(
+        // TODO
+        // Add throttling key for other master task as well.
+        // Will be added as part of upcoming PRs,
+        // Right now just added "put-mapping" for ref only.
+        new HashSet<>(Arrays.asList("put-mapping"))
+    );
+
+    private final int DEFAULT_THRESHOLD_VALUE = -1; // Disabled throttling
+    private final MasterService masterService;
+    private final MasterTaskThrottlerListener masterTaskThrottlerListener;
+
+    private final ConcurrentMap<String, Long> tasksCount;
+    private final ConcurrentMap<String, Long> tasksThreshold;
+
+    public MasterTaskThrottler(
+        final ClusterSettings clusterSettings,
+        final MasterService masterService,
+        final MasterTaskThrottlerListener masterTaskThrottlerListener
+    ) {
+        clusterSettings.addSettingsUpdateConsumer(THRESHOLD_SETTINGS, this::updateSetting, this::validateSetting);
+        this.masterService = masterService;
+        this.masterTaskThrottlerListener = masterTaskThrottlerListener;
+        tasksCount = new ConcurrentHashMap<>(128); // setting initial capacity so each task will lend in different segment
+        tasksThreshold = new ConcurrentHashMap<>(128); // setting initial capacity so each task will lend in different segment
+    }
+
+    public void validateSetting(final Settings settings) {
+        /**
+         * TODO: Change the version number of check as per version in which this change will be merged.
+         */
+        if (masterService.state().nodes().getMinNodeVersion().compareTo(Version.V_3_0_0) < 0) {
+            throw new IllegalArgumentException("All the nodes in cluster should be on version later than or equal to 3.0.0");
+        }
+        Map<String, Settings> groups = settings.getAsGroups();
+        for (String key : groups.keySet()) {
+            if (!CONFIGURED_TASK_FOR_THROTTLING.contains(key)) {
+                throw new IllegalArgumentException("Master task throttling is not configured for given task type: " + key);
+            }
+            int threshold = groups.get(key).getAsInt("value", DEFAULT_THRESHOLD_VALUE);
+            if (threshold < DEFAULT_THRESHOLD_VALUE) {
+                throw new IllegalArgumentException("Provide positive integer for limit or -1 for disabling throttling");
+            }
+        }
+    }
+
+    public void updateSetting(final Settings settings) {
```

Comment:
Can we make this package private please?

Reply:
Ack

Reply:
Ack